cmd/poller/asup.go

```diff
@@ -0,0 +1,269 @@
+package main
+
+import (
+	"encoding/json"
+	"goharvest2/cmd/harvest/version"
+	"goharvest2/pkg/conf"
+	"goharvest2/pkg/matrix"
+	"goharvest2/pkg/tree/node"
+	"io/ioutil"
+	"os/exec"
+	"strconv"
+	"strings"
+)
+
+type asupMessage struct {
+	Target   *targetInfo
+	Nodes    *instanceInfo
+	Volumes  *instanceInfo
+	Platform *platformInfo
+	Harvest  *harvestInfo
+}
+
+type targetInfo struct {
+	Version string
+	Model   string
+	Serial  string
+	Ping    float64
+}
+
+type instanceInfo struct {
+	Count      int64
+	DataPoints int64
+	PollTime   int64
+	ApiTime    int64
+	ParseTime  int64
+	PluginTime int64
+}
+
+type platformInfo struct {
+	OS       string
+	Arch     string
+	MemoryKb uint64
+	CPUs     uint8
+}
+
+type harvestInfo struct {
+	UUID        string
+	Version     string
+	Release     string
+	Commit      string
+	BuildDate   string
+	NumClusters uint8
+}
+
+func (p *Poller) doAsupMessage() (*matrix.Matrix, error) {
+
+	var (
+		msg *asupMessage
+		err error
+	)
+
+	if msg, err = p.buildAsupMessage(); err != nil {
+		logger.Error().Err(err).Msg("failed to build ASUP message")
+		return nil, err
+	}
+
+	if err = p.sendAsupMessage(msg); err != nil {
+		logger.Error().Err(err).Msg("failed to send ASUP message")
+		return nil, err
+	}
+
+	return nil, nil
+
+}
+
+// TODO: This function will be used to invoke harvest-asup(private repo)
+func (p *Poller) sendAsupMessage(msg *asupMessage) error {
+	logger.Info().Msgf("ASUP generatation started")
+
+	file, _ := json.MarshalIndent(msg, "", " ")
+	_ = ioutil.WriteFile("test.json", file, 0644)
+
+	logger.Info().Msgf("ASUP HH generation done !!!")
+	return nil
+}
+
+func (p *Poller) buildAsupMessage() (*asupMessage, error) {
+
+	var (
+		msg  *asupMessage
+		arch string
+		cpus uint8
+	)
+
+	// @DEBUG all log messages are info or higher, only for development/debugging
+	logger.Info().Msg("building ASUP message")
+
+	msg = new(asupMessage)
+
+	// add harvest release info
+	msg.Harvest = new(harvestInfo)
+	msg.Harvest.UUID = harvestUUID.String()
+	msg.Harvest.Version = version.VERSION
+	msg.Harvest.Release = version.Release
+	msg.Harvest.Commit = version.Commit
+	msg.Harvest.BuildDate = version.BuildDate
+	msg.Harvest.NumClusters = getNumClusters(p.options.Config)
+
+	// add info about platform (where Harvest is running)
+	msg.Platform = new(platformInfo)
+	arch, cpus = getCPUInfo()
+	msg.Platform.Arch = arch
+	msg.Platform.CPUs = cpus
+	msg.Platform.MemoryKb = getRamSize()
+	msg.Platform.OS = getOSName()
+
+	// info about ONTAP host and instances
+	msg.Target, msg.Nodes, msg.Volumes = p.getInstanceInfo()
+
+	return msg, nil
+}
+
+func (p *Poller) getInstanceInfo() (*targetInfo, *instanceInfo, *instanceInfo) {
+	target := new(targetInfo)
+	nodes := new(instanceInfo)
+	vols := new(instanceInfo)
+
+	// get ping value from poller metadata
+	target.Ping, _ = p.status.LazyGetValueFloat64("ping", "host")
+
+	// scan collectors
+
+	for _, c := range p.collectors {
+
+		if c.GetName() == "ZapiPerf" {
```

Comment:
should this be `Zapi` instead of `ZapiPerf`

Reply:
done